OpenRA.Game/Activities/Activity.cs

```diff
@@ -9,50 +9,218 @@
+			}
+
+			set
+			{
+				if (value == this || value == Parent || value == nextActivity)
+					childActivity = null;
+				else
+				{
+					childActivity = value;
+
+					if (childActivity != null)
+						childActivity.Parent = this;
+				}
+			}
+		}
+
+		protected Activity nextActivity;
+		public virtual Activity NextActivity
+		{
+			get
+			{
+				return nextActivity != null ? nextActivity : Parent;
+			}
+
+			set
+			{
+				if (value == this || value == Parent || (value != null && value.Parent == this))
+					nextActivity = null;
+				else
+				{
+					nextActivity = value;
+
+					if (nextActivity != null)
+						nextActivity.Parent = Parent;
+				}
+			}
+		}
+
 		public bool IsInterruptible { get; protected set; }
-		protected bool IsCanceled { get; private set; }
+		protected bool IsCanceled { get { return State == ActivityState.Canceled; } }
 
 		public Activity()
 		{
 			IsInterruptible = true;
 		}
 
+		public Activity RealTick(Actor self)
+		{
+			if (State == ActivityState.Done && Game.Settings.Debug.StrictActivityChecking)
+				throw new InvalidOperationException("Actor {0} attempted to tick activity {1} after it had already completed.".F(self, this.GetType()));
+
+			if (State == ActivityState.Queued)
+			{
+				OnFirstRun(self);
+				State = ActivityState.Active;
+			}
+
+			var ret = Tick(self);
+			if (ret != this)
+			{
+				// Make sure that the ChildActivity pointer is moved forwards as the child queue of a CompositeActivity advances
+				if (ret != null && ret.parent == this)
+					childActivity = ret;
+				else
+				{
+					if (State != ActivityState.Canceled)
+						State = ActivityState.Done;
+
+					OnLastRun(self);
+				}
+			}
+
+			return ret;
+		}
+
 		public abstract Activity Tick(Actor self);
 
+		/// <summary>
+		/// Runs once before the first Tick() execution.
+		/// </summary>
+		protected virtual void OnFirstRun(Actor self) { }
+
+		/// <summary>
+		/// Runs once after the last Tick() execution.
+		/// </summary>
+		protected virtual void OnLastRun(Actor self) { }
+
 		public virtual bool Cancel(Actor self)
 		{
 			if (!IsInterruptible)
 				return false;
 
-			IsCanceled = true;
+			if (ChildActivity != null && !ChildActivity.Cancel(self))
+				return false;
+
+			State = ActivityState.Canceled;
 			NextActivity = null;
+			ChildActivity = null;
 
 			return true;
 		}
 
 		public virtual void Queue(Activity activity)
 		{
-			if (NextActivity != null)
-				NextActivity.Queue(activity);
+			if (nextActivity != null)
+				nextActivity.Queue(activity);
 			else
 				NextActivity = activity;
 		}
 
+		public virtual void QueueChild(Activity activity)
+		{
+			if (ChildActivity != null)
+				ChildActivity.Queue(activity);
+			else
+				ChildActivity = activity;
+		}
+
+		protected void PrintActivityTree(Activity origin = null, int level = 0)
+		{
+			if (origin == null)
+				Root.PrintActivityTree(this);
+			else
+			{
+				Console.Write("".PadLeft(level * 2));
```

Comment:
No reason, except I didn't know that this ctor existed, so I went with the first suggestion Google threw at me. Will change, as that is exactly what I was looking for.